content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,63 @@
+
+Before adding or removing an item from the LRU cache, one needs to find first if an item already exists. Using our initial example, let us access the third item from our cache i.e. 2. We first do a search to see whether 2 is available. Yes, it is available. This means that we will have to shift item 2 to the front of our Cache or list. Our new cache order will be 2,8,1,3. 2 becomes the Most Recently Accessed item. With this, we notice that when we want to access an item, it shifts to the front of the list. So, when we want to remove an item, we pick the one that is at the end of the list and remove it from our cache. The item at the end of our list is usually the Least Recently Used item.
+
+### Data Structures Used to Build An Efficient LRU Cache Algorithm.
+
+From the example above, we performed many operations while building the LRU cache. These operations include
+
+1. Searching:- We want to get item 2 from our list. First, we need to check on our entire list to find whether it is there or not. If the element is present, then we have to reorder the list. We move 2 to the beginning of the list. In this case, the time complexity for re-ordering will be O(n).
+2. Let's say we have a scenario where we need to add e.g. 8 into our list. Our list, in this case, is still 1,2,3,4. Searching through the list, we can't find an 8. The next thing we need to do since our cache size is 4, we remove 4 from the list. Once that is done, we then shift 1,2, and 3 to the left. We are now left with an empty slot at the beginning of the list to add the 8. The time complexity for doing the shifting is O(n).
+
+We can make optimize the shifting by eliminating it. To do so we use a double liked list. Why?
+
+Still using the example 1,2,3,4. We needed to add 8. Since it wasn't in our list, what we did was remove the oldest element 4 and shift the entire list items. In a doubly-linked list, all you need to do is to remove the rear and update the rear element to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation is 0(1) time complexity.
+
+We can also optimize searching. How?
+
+Still using our previous example of cache size 4, with 1,2,3,4. searching for 8. We can optimize our search by using a hash table. With a Hash, you will have a key and a value pair. The Key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.
+
+[Hash Table & Double Linked List Diagram](https://leetcode.com/problems/ddl_hash.png/)
+

```

Comment:
Please download the image and add a link to the source. This way, if the source fails, you still have the image on your article.